contrib/db_pools/lib/src/database.rs

```diff
@@ -0,0 +1,137 @@
+use rocket::fairing::{Info, Kind};
+use rocket::futures::future::BoxFuture;
+use rocket::{Build, Ignite, Rocket, Sentinel};
+
+use crate::Pool;
+
+/// Trait implemented to define a database connection pool.
+pub trait Database: Sized + Send + Sync + 'static {
+    /// The name of this connection pool in the configuration.
+    const NAME: &'static str;
+
+    /// The underlying connection type returned by this pool.
+    /// Must implement [`Pool`].
+    type Pool: Pool;
+
+    /// Returns a fairing that attaches this connection pool to the server.
+    fn fairing() -> Fairing<Self>;
+
+    /// Direct shared access to the underlying database pool
+    fn pool(&self) -> &Self::Pool;
+
+    /// get().await returns a connection from the pool (or an error)
+    fn get(&self) -> BoxFuture<'_, Result<Connection<Self>, <Self::Pool as Pool>::GetError>> {
```

Comment:
A previous design of this did use `async_trait`, but you expressed a wish to avoid `async_trait`. Although maybe I misunderstood where/why you wanted to avoid it?

Reply:
Okay... in that case, should the derive generate an inherent `async` method, an `#[async_trait]` trait method, or both?